### PR TITLE
WIP feat(router): adds routing based on URL templates

### PR DESCRIPTION
**Pull request by hypeJunction**
Implements a new approach to routing using URL templates.
Uses Symfony Routing Component for URL matching and URL generation
based on predefined URL templates.
Adds elgg_register_route() and elgg_unregister_route()

TODOs:

- [ ] Deprecate `elgg_register_page_handler()`
- [ ] Stop accepting buffer output from `route` hook - must return `ResponseBuilder`
- [ ] Stop allowing route modifications from `route` hook
- [ ] Add route caching
- [ ] More thorough documentation
- [ ] Tests

### CODE REVIEW

**Comment by hypeJunction** [comment]
Here is the first stab at it. Let me know your thoughts. I quite like it so far.
We will need to think of a better way to register/cache routes. We probably need to add a route registration service, which will then trigger a hook to filter/reorder routes, and a system event that will register the routes.

File: docs/guides/upgrading.rst [hunk 1]
```diff
@@ -321,6 +321,7 @@ Miscellaneous API changes
  * The generic comment save action no longer sends the notification directly, this has been offloaded to the notification system.
  * The script ``engine/start.php`` is removed.
  * The functions ``set_config``, ``unset_config`` and ``get_config`` have been deprecated and replaced by ``elgg_set_config``, ``elgg_remove_config`` and ``elgg_get_config``. 
+ * ``'route','<identifier>'`` hook can no longer be used to modify routes. Plugins should use ``'route:rewrite','<identifier>'`` hook with a handler registered before the ``'init','system'`` event is fired.
```
Comment (hypeJunction) [inline docs/guides/upgrading.rst]: This not yet true. I figured we can match by path, so it wasn't necessary to break anything.

File: mod/blog/elgg-plugin.php [hunk 2]
```diff
@@ -14,6 +14,57 @@
...
+			'requirements' => [
+				'revision' => '\d+',
+			],
+		],
+		'blog_group' => [
+			'path' => '/blog/group/{group_guid}/{subpage?}/{lower?}/{upper?}',
+			'resource' => 'blog/group',
+			'defaults' => [
+				'subpage' => 'all',
+				'lower' => '',
```
Comment (hypeJunction) [inline mod/blog/elgg-plugin.php]: not needed

File: mod/blog/elgg-plugin.php [hunk 3]
```diff
@@ -14,6 +14,57 @@
...
+		],
+		'blog_archive' => [
+			'path' => '/blog/archive/{username?}/{lower?}/{upper?}',
+			'resource' => 'blog/archive',
+			'requirements' => [
+				'lower' => '\d+',
+				'upper' => '\d+',
+			],
+		],
+		'blog_view' => [
```
Comment (hypeJunction) [inline mod/blog/elgg-plugin.php]: For `view` we probably need some system-wide approach, so we can resolve URLs to entities.

File: mod/blog/elgg-plugin.php [hunk 4]
```diff
@@ -14,6 +14,57 @@
...
+			'path' => '/blog/view/{guid}/{title?}',
+			'resource' => 'blog/view',
+		],
+		'blog_add' => [
+			'path' => '/blog/add/{guid?}',
+			'resource' => 'blog/add',
+		],
+		'blog_edit' => [
```
Comment (hypeJunction) [inline mod/blog/elgg-plugin.php]: For `edit` we need some system-wide approach, so we can have `/edit` as the primary route, which then renders a resource that has a predictable name `/resources/edit/<type>/<subtype>`

File: mod/blog/lib/blog.php [hunk 5]
```diff
@@ -81,9 +81,14 @@ function blog_get_page_content_archive($owner_guid, $lower = 0, $upper = 0) {
...
 	$crumbs_title = $owner->name;
 	if (elgg_instanceof($owner, 'user')) {
-		$url = "blog/owner/{$owner->username}";
+		$url = elgg_generate_url('blog_owner', [
+			'username' => $owner->username,
+		]);
 	} else {
-		$url = "blog/group/$owner->guid/all";
+		$url = elgg_generate_url('blog_group', [
+			'guid' => $owner->guid,
```
Comment (hypeJunction) [inline mod/blog/lib/blog.php]: `group_guid`

File: mod/blog/views/default/resources/blog/add.php [hunk 6]
```diff
@@ -2,7 +2,12 @@
 
 elgg_gatekeeper();
 
-$page_type = elgg_extract('page_type', $vars);
+elgg_load_library('elgg:blog');
+
+// push all blogs breadcrumb
+elgg_push_breadcrumb(elgg_echo('blog:blogs'), elgg_generate_url('blog_all'));
+
+$page_type = 'add';
```
Comment (hypeJunction) [inline mod/blog/views/default/resources/blog/add.php]: Add to `$vars`

File: mod/blog/views/default/resources/blog/all.php [hunk 7]
```diff
@@ -1,6 +1,11 @@
 <?php
 
-$page_type = elgg_extract('page_type', $vars);
+$page_type = 'all';
```
Comment (hypeJunction) [inline mod/blog/views/default/resources/blog/all.php]: Add to `$vars`

File: mod/blog/views/default/resources/blog/archive.php [hunk 8]
```diff
@@ -1,6 +1,11 @@
 <?php
 
-$page_type = elgg_extract('page_type', $vars);
+elgg_load_library('elgg:blog');
+
+// push all blogs breadcrumb
+elgg_push_breadcrumb(elgg_echo('blog:blogs'), elgg_generate_url('blog_all'));
+
+$page_type = 'archive';
```
Comment (hypeJunction) [inline mod/blog/views/default/resources/blog/archive.php]: Add to `$vars`

File: mod/blog/views/default/resources/blog/edit.php [hunk 9]
```diff
@@ -2,7 +2,12 @@
 
 elgg_gatekeeper();
 
-$page_type = elgg_extract('page_type', $vars);
+elgg_load_library('elgg:blog');
+
+// push all blogs breadcrumb
+elgg_push_breadcrumb(elgg_echo('blog:blogs'), elgg_generate_url('blog_all'));
+
+$page_type = 'edit';
```
Comment (hypeJunction) [inline mod/blog/views/default/resources/blog/edit.php]: Add to `$vars`

File: mod/blog/views/default/resources/blog/friends.php [hunk 10]
```diff
@@ -1,6 +1,11 @@
 <?php
 
-$page_type = elgg_extract('page_type', $vars);
+elgg_load_library('elgg:blog');
+
+// push all blogs breadcrumb
+elgg_push_breadcrumb(elgg_echo('blog:blogs'), elgg_generate_url('blog_all'));
+
+$page_type = 'friends';
```
Comment (hypeJunction) [inline mod/blog/views/default/resources/blog/friends.php]: Add to `$vars`

File: mod/blog/views/default/resources/blog/group.php [hunk 11]
```diff
@@ -1,7 +1,12 @@
 <?php
 
+elgg_load_library('elgg:blog');
+
+// push all blogs breadcrumb
+elgg_push_breadcrumb(elgg_echo('blog:blogs'), elgg_generate_url('blog_all'));
+
 $subpage = elgg_extract('subpage', $vars);
-$page_type = elgg_extract('page_type', $vars);
+$page_type = 'group';
```
Comment (hypeJunction) [inline mod/blog/views/default/resources/blog/group.php]: Add to `$vars`

File: mod/blog/views/default/resources/blog/owner.php [hunk 12]
```diff
@@ -1,6 +1,11 @@
 <?php
 
-$page_type = elgg_extract('page_type', $vars);
+elgg_load_library('elgg:blog');
+
+// push all blogs breadcrumb
+elgg_push_breadcrumb(elgg_echo('blog:blogs'), elgg_generate_url('blog_all'));
+
+$page_type = 'owner';
```
Comment (hypeJunction) [inline mod/blog/views/default/resources/blog/owner.php]: Add to `$vars`

File: mod/blog/views/default/resources/blog/view.php [hunk 13]
```diff
@@ -1,6 +1,11 @@
 <?php
 
-$page_type = elgg_extract('page_type', $vars);
+elgg_load_library('elgg:blog');
+
+// push all blogs breadcrumb
+elgg_push_breadcrumb(elgg_echo('blog:blogs'), elgg_generate_url('blog_all'));
+
+$page_type = 'view';
```
Comment (hypeJunction) [inline mod/blog/views/default/resources/blog/view.php]: Add to `$vars`

File: mod/blog/elgg-plugin.php [hunk 14]
```diff
@@ -14,6 +14,57 @@
 		'blog/auto_save_revision' => [],
 		'blog/delete' => [],
 	],
+	'routes' => [
```
Comment (hypeJunction) [inline mod/blog/elgg-plugin.php]: We can add an opiton to make the route public and automatically register it for walled garden.

File: mod/blog/elgg-plugin.php (hunk 3 above)
Comment (hypeJunction) [inline mod/blog/elgg-plugin.php]: Maybe we could do `view:<type>:<subtype>` as the route name

File: mod/blog/elgg-plugin.php [hunk 15]
```diff
@@ -14,6 +14,57 @@
...
+		'blog_owner' => [
```
Comment (hypeJunction) [inline mod/blog/elgg-plugin.php]: Would `blog:owner` be more Elgg'y?

**Comment by jdalsem** [comment]
Wow! This is really nice...

**Pull request closed without merging**